synapse/rest/media/v1/preview_url_resource.py

```diff
@@ -717,32 +717,28 @@ def decode_and_calc_og(
         parser = etree.HTMLParser(recover=True, encoding=request_encoding)
     except LookupError:
         # blindly consider the encoding as utf-8.
-        try:
-            parser = etree.HTMLParser(recover=True, encoding="utf-8")
-        except Exception as e:
-            logger.warning("Unable to create fallback HTML parser: %s" % (e,))
-            return {}
+        parser = etree.HTMLParser(recover=True, encoding="utf-8")
     except Exception as e:
         logger.warning("Unable to create HTML parser: %s" % (e,))
         return {}
 
     # Attempt to parse the body. If this fails, log and return no metadata.
     try:
         tree = etree.fromstring(body, parser)
+        return _calc_og(tree, media_uri)
     except UnicodeDecodeError:
         # blindly try decoding the body as utf-8, which seems to fix
         # the charset mismatches on https://google.com
         try:
             tree = etree.fromstring(body.decode("utf-8", "ignore"), parser)
+            return _calc_og(tree, media_uri)
         except Exception as e:
             logger.warning("Failed to parse HTML body as UTF-8: %s" % (e,))
             return {}
     except Exception as e:
         logger.warning("Failed to parse HTML body: %s" % (e,))
         return {}
 
-    return _calc_og(tree, media_uri)
```

Comment:
I'm not sure it makes sense to do this? As a) if it fails with `UnicodeDecodeError` we'll pointlessly retry and b) the log messages won't really be accurate.

Reply:
I'm not sure what this is referring to. This code changed in the final commit (when I added tests).

Reply:
I mean, why are we moving them inside the try/except blocks?

Reply:
They've always been inside the try-except block. An early commit on this branch moved them out, then I realized an error can be thrown from them (and that's why they were inside of them) so I moved them back.

Reply:
Ooooh, right. I'm still tempted to move `return _calc_og(tree, media_uri)`  out and wrap it in a separate try/catch, but :shrug: up to you

Reply:
If the `_calc_og` raises, then we need to create the new `tree` with the `body.decode("utf-8", "ignore")` bit, I'll try to make this a bit more readable, but I'm not sure...

Reply:
Oh, I really wasn't expecting it to fail with further decode errors? So `etree.fromstring` *and* `_calc_og` may run into decoding issues? Is that because `etree.fromstring` doesn't fully parse it and then `_calc_og` runs into it? Or?

I feel like I'm being distinctly unhelpful here, let's discuss this when you're around so we can get this over the line without another set  of back and forths

Reply:
The `UnicodeDecodeError` comes out of `_calc_og` when you start iterating the tree, *not* when you call `etree.fromstring`. Maybe that's where the confusion lies?

Reply:
Aaaah, yeah OK. I think you can just ignore me then